File: src/test/java/com/example/webfluxdemo/services/PersonalizedGreetingServiceTest.java
package com.example.webfluxdemo.services;

import com.example.webfluxdemo.adapters.TranslationAdapter;
import com.example.webfluxdemo.adapters.UserAdapter;
import com.example.webfluxdemo.model.PersonalizedGreeting;
import com.example.webfluxdemo.model.User;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class PersonalizedGreetingServiceTest {

    private final UserAdapter userAdapter = mock(UserAdapter.class);
    private final TranslationAdapter translationAdapter = mock(TranslationAdapter.class);

    private final PersonalizedGreetingService personalizedGreetingService = new PersonalizedGreetingService(
            userAdapter,
            translationAdapter
    );

    @Test
    void testPersonalizeGreeting() {
        when(userAdapter.getUser(123)).thenReturn(new User("Francois", "fr-FR"));
        when(translationAdapter.getTranslation("hello", "fr-FR")).thenReturn("Bonjour");

        PersonalizedGreeting actual = personalizedGreetingService.personalizeGreeting(123);

        PersonalizedGreeting expected = new PersonalizedGreeting("Bonjour", "Francois");
        assertThat(actual).isEqualTo(expected);
    }
}

File: src/test/java/com/example/webfluxdemo/services/PersonalizedGreetingServiceTest.kt
package com.example.webfluxdemo.services

import com.example.webfluxdemo.adapters.TranslationAdapter
import com.example.webfluxdemo.adapters.UserAdapter
import com.example.webfluxdemo.model.PersonalizedGreeting
import com.example.webfluxdemo.model.User
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.AssertionsForClassTypes.assertThat
import org.junit.jupiter.api.Test
import reactor.core.publisher.Mono

internal class PersonalizedGreetingServiceTest {

    private val userAdapter = mockk<UserAdapter>()
    private val translationAdapter = mockk<TranslationAdapter>()

    private val personalizedGreetingService = PersonalizedGreetingService(
            userAdapter,
            translationAdapter
    )

    @Test
    fun testPersonalizeGreeting() {
        every { userAdapter.getUser(123) } returns
                Mono.just(User("Francois", "fr-FR"))

        every { translationAdapter.getTranslation("hello", "fr-FR") } returns
                Mono.just("Bonjour")

        val actual = personalizedGreetingService.personalizeGreeting(123).block()
        val expected = PersonalizedGreeting("Bonjour", "Francois")
        assertThat(actual).isEqualTo(expected)
    }
}